"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const iterablex_1 = require("./iterablex");
const returniterator_1 = require("../internal/returniterator");
class CatchIterable extends iterablex_1.IterableX {
    constructor(source) {
        super();
        this._source = source;
    }
    *[Symbol.iterator]() {
        let error = null, hasError = false;
        for (let source of this._source) {
            const it = source[Symbol.iterator]();
            error = null;
            hasError = false;
            while (1) {
                let c = {};
                try {
                    const { done, value } = it.next();
                    if (done) {
                        returniterator_1.returnIterator(it);
                        break;
                    }
                    c = value;
                }
                catch (e) {
                    error = e;
                    hasError = true;
                    returniterator_1.returnIterator(it);
                    break;
                }
                yield c;
            }
            if (!hasError) {
                break;
            }
        }
        if (hasError) {
            throw error;
        }
    }
}
exports.CatchIterable = CatchIterable;
/**
 * Creates a sequence by concatenating source sequences until a source sequence completes successfully.
 * @param {Iterable<Iterable<TSource>>} source Source sequences.
 * @return {Iterable<TSource>} Sequence that continues to concatenate source sequences while errors occur.
 */
function _catchAll(source) {
    return new CatchIterable(source);
}
exports._catchAll = _catchAll;
/**
 * Creates a sequence by concatenating source sequences until a source sequence completes successfully.
 * @param {Iterable<TSource>} source The first source.
 * @param {...Iterable<TSource>} args The rest of the sequence that continues to concatenate source sequences while errors occur.
 */
function _catch(source, ...args) {
    return new CatchIterable([source, ...args]);
}
exports._catch = _catch;
/**
 * Creates a sequence by concatenating source sequences until a source sequence completes successfully.
 * @param {...Iterable<TSource>} source Sequence that continues to concatenate source sequences while errors occur.
 * @return {Iterable<TSource>} Sequence that continues to concatenate source sequences while errors occur.
 */
function _catchStatic(...source) {
    return new CatchIterable(source);
}
exports._catchStatic = _catchStatic;

//# sourceMappingURL=catch.js.map
